Tensor spaces, second try

kohr-h opened the pull request:
Copied from #753

TODOs:
- [x] Implement basic structures for tensors and tensor spaces
- [x] Replicate functionality of `NtuplesBase`, `FnBase`, `NumpyNtuples` and `NumpyFn` in the corresponding tensor classes
- [x] Remove `Ntuples` and related stuff
- [x] Add former `Ntuples` and the likes as special case to tensor classes
- [x] Change names with "tensor" to something more lightweight
- [x] Find place for ~~`MatVecOperator`~~`MatrixOperator`, see #584. Solved in #911 
- [x] Improve pretty-printing of elements
- [x] Fix broken downstream code (`lp_discr`, ...)
- [x] Change code that was restricted to 1D data containers
- [x] Port unit tests
- [x] Write some doctests
- [x] Write more doctests
- [x] Remove `TensorSet` and others (see #860) duck-type space properties
- [x] Check documentation for outdated information, e.g. regarding 1D data layout (Update: can't find anything explicit)
- [x] ~~Use notion "rank" instead of "number of dimensions" to better distinguish between `space.ndim` and `space.size`~~ Not doing this
- [x] Implement indexing of `TensorSpace`. ~~Done, but kind of wrong, see [here](https://github.com/odlgroup/odl/pull/861#issuecomment-275454215) and [here](https://github.com/odlgroup/odl/pull/861#issuecomment-275540310)~~ Done properly now
- [x] ~~Write extra documentation on tensors~~ Should be part of #908
- [x] Port `show` methods (1D from `Ntuples`, 2D from `DiscreteLp`)
- [ ] Add unit tests for unusual data types and the likes
- [x] Simplify doctests of tensor spaces
- [x] Fix `FunctionSpaceElement` in 1D (again)
- [x] Implement tensor-valued functions, see #908 
- [x] Fix interpolation from array of points, related to #908 
- [x] Cleanup (old stuff, commit history, ...)
- [ ] Fix large-scale tests
- [x] Make test with power function more general
- [ ] Add more tests for product space multi-indexing
- [ ] Implement indexing of `FunctionSpace` as 
  - [ ] indexing the components of the output (i.e. raise an error for scalar-valued functions) in `__getitem__`
  - [ ] by axis of the domain using a `byaxis` attribute
- [ ] Implement indexing of `DiscreteLp` as in `FunctionSpace`
- [x] Implement `__array_ufunc__` interface on `Tensor` and `NumpyTensor` (the latter only differs in weight propagation)
- [ ] Implement `__array_ufunc__` interface on `DiscreteLpElement`
- [ ] Write more extensive tests for `__array_ufunc__` intefaces (bad input etc.)

kohr-h commented:
> Regarding this topic: I have implemented a simple indexing capability for spaces, but it is nowhere near as powerful and flexible as the lazy approach of simply indexing the array and appropriately wrap whatever comes out.
> So the earlier suggestion
> 
> arr = self.data[indices]
> return self.space[indices].element(arr)
> 
> is much much less powerful than lazy indexing, if we don't put great amounts of work into it.
> 
> So I'd say the space indexing can be used for simple things like
> 
> >>> space = odl.rn((2, 3, 4))
> >>> space[[0, 2]]
> rn((2, 4))
> >>> space[0]
> rn(2)
> 
> but not for indexing of elements.

I realized that indexing like this is not really helpful since it's not compatible with power spaces. I'll think about it a bit more.

adler-j commented:
Is this in a state where it needs a review? If so I guess I need to get at it.

kohr-h commented:
> Is this in a state where it needs a review? If so I guess I need to get at it.

Yes it is. I deliberately didn't mention it so nobody burns time at it that they don't have :-)

adler-j commented:
Could you do the rebase from master first? It seems rather big.

kohr-h commented:
> Could you do the rebase from master first? It seems rather big.

Yes, I've also spotted some glitches that I want to fix along the way.

kohr-h commented:
Rebased

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        tensor_space :
+            constructor for tensor spaces of arbitrary scalar
+            data type
+
```
Extra blankrow

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        >>> space
+        rn((2, 3))
+
+        >>> space = NumpyTensorSpace((2, 3), dtype='complex64')
```
Note earlier comment on shape. Should be integral in general unless having several shows something.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        dist_using_inner = bool(kwargs.pop('dist_using_inner', False))
+
+        if (not self.is_numeric and
+                any(x is not None for x in (dist, norm, inner, weighting))):
```
x is a bad name

kohr-h replied:
I guess it's okay in a generator expression?

adler-j replied:
Well sure it is OK, but it is somewhat missleading regarding what it is, can't think of anything better atm though.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+            raise ValueError('cannot use any of `dist`, `norm` or `inner` '
+                             'for exponent != 2')
+        # Check validity of option combination (3 or 4 out of 4 must be None)
+        if sum(x is None for x in (dist, norm, inner, weighting)) < 3:
```
`np.count_nonzero((dist, norm, inner, weighting)) < 3` also works. is it more readable?

kohr-h replied:
I'm not really fond of code that relies on `None` evaluating to `False`. That's like statements
```python
impl = str(impl).lower() or 'none'
```
instead of 
```python
impl = 'none' if impl is None else str(impl).lower()
```
I find the latter more readable and more explicit.

adler-j replied:
While I agree, this code simply pushes the problem to `False` evaluating to `0`, which while more simple is not much better.

kohr-h replied:
Actually the simplest and most correct version is
```python
if (dist, norm, inner, weighting).count(None) < 3:
    # ...
```

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+
+            If ``inp`` is a `numpy.ndarray` of the same shape and data
+            type as this space, the array is wrapped, not copied.
+            Other array-like objects are copied.
```
necessarily copied, or something like that

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        all_dtypes = []
+        for lst in np.sctypes.values():
+            all_dtypes.extend(set(lst))
+        dtypes = [np.dtype(dtype) for dtype in all_dtypes]
```
Rather have two loops than a loop then a list comprehension, e.g. either

```
all_dtypes = sum(np.sctypes.values(), [])
dtypes = set(np.dtype(dtype) for dtype in all_dtypes)
```

or

```
all_dtypes = []
for lst in np.sctypes.values():
    for dtype in lst:
        all_dtypes.append(np.dtype(dtype))
```

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        )
+        >>> out
+        cn((2, 3)).element(
+            [[0j, (-1+0j), 2j],
```
weird spacing compared to the above

kohr-h replied:
This is the correct one according to the implementation. That's up to debate. Just weird that the other one is accepted, too. Probably due to the `NORMALIZE_WHITESPACE` option of doctest.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+                     self.weighting))
+
+    def __getitem__(self, indices):
+        """Return ``self[indices]``."""
```
This could use a doctest

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+
+    def __init__(self, space, data):
+        """Initialize a new instance."""
+        assert isinstance(space, NumpyTensorSpace)
```
Again, I think  we should remove this asserts for performance reasons since they should only capture developer errors.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        else:
+            return np.array_equal(self.data, other.data)
+
+    def __hash__(self):
```
We do not hash mutable objects

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        """
+        return self.space.element(self.data.copy())
+
+    __copy__ = copy
```
See #903

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+            return self.space.zero()
+        elif self.space.is_complex_space:
+            # Definitely non-contiguous
+            real_space = self.space.astype(self.space.real_dtype, order='A')
```
`self.space.real_space`?

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+            return self
+        elif self.space.is_complex_space:
+            # Definitely non-contiguous
+            real_space = self.space.astype(self.space.real_dtype, order='A')
```
`self.space.real_space`?

kohr-h replied:
The point here is that we force `order='A'` instead of inheriting that property from the space. As the comment says (hopefully it's true) we know the result won't be contiguous.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+    def __ipow__(self, other):
+        """Return ``self **= other``."""
+        try:
+            if other == int(other):
```
prefer try else
also what if this if statement fails?

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        dot = np.vdot  # slowest alternative
+    # x2 as first argument because we want linearity in x1
+    return dot(x2.data.ravel(order=x2.order),
+               x1.data.ravel(order=x1.order))
```
This can fail if x1 and x2 have different orders

kohr-h replied:
How? I don't think `dot` cares.

adler-j replied:
Well it doesn't but ravel does. See e.g.

```python
>>> x1 = np.array([[1, 2],
...                [3, 4]], order='F')
>>> x2 = np.array([[1, 2],
...                [3, 4]], order='C')
>>> np.dot(x2.ravel(order='C'),
...        x1.ravel(order='F'))
29
>>> np.vdot(x1, x2)
30
```

adler-j replied:
In the above, we expect 1 + 4+ 9 + 16 = 30, hence your method gives the wrong answer.

kohr-h replied:
Okay, that's not failure, but wrong answer. And agreed, that can happen, but not really because `x1` and `x2` are checked to be in the same space before. Anyway, they should be flattened with the same order just in case.

adler-j replied:
Well the orders are not enforced in that sense, i.e. spaces with order='A' could have both.

The easiest suggestion (and likely fastest) is to simply ravel both of them according to the same scheme, i.e.

```python
dot(x2.data.ravel(order=x1.order),
    x1.data.ravel(order=x1.order))
```

that could cause a copy, but is always correct and does not cause a copy when not needed. Another option is to make sure this actually happens on the multidimensional arrays, e.g. using `einsum`.

kohr-h replied:
Last time I tested `einsum` it was bloody slow, but that was in the case of matrix times tensor along one axis.
Simple test:
```python
>>> a = b = np.ones(int(1e8))
>>> np.dot(a, b)
100000000.0
>>> np.einsum('i,i', a, b)
100000000.0
>>> np.tensordot(a, b, axes=(0, 0))
array(100000000.0)

>>> %timeit np.dot(a, b)
10 loops, best of 3: 52.5 ms per loop
>>> %timeit np.einsum('i,i', a, b)
10 loops, best of 3: 50.1 ms per loop
>>> %timeit np.tensordot(a, b, axes=(0, 0))
10 loops, best of 3: 52.5 ms per loop
```

adler-j replied:
Tensor dot seems like the most readable of those then?

kohr-h replied:
Yes, I agree. Will use it.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+               x1.data.ravel(order=x1.order))
+
+
+# TODO: implement intermediate weighting schemes with arrays that are
```
Don't now, but make an issue.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+        super().__init__(array, impl='numpy', exponent=exponent,
+                         dist_using_inner=dist_using_inner)
+
+    def __hash__(self):
```
hash but no `__eq__`?

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+
+    """Weighting of a `NumpyTensorSpace` with constant 1."""
+
+    # Implement singleton pattern for efficiency in the default case
```
Wait a minute. This has to fail in weird ways if several spaces are created, no?

adler-j replied:
Edit, no it does not since this does not have a specific domain. Anyway, please add a few words on that.

kohr-h replied:
I've hardly touched this code since it came about earlier. It's worked ever since. If you remember your original reasons you're welcome to share them :-) I'll try to come up with something in the meanwhile.

adler-j commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,2336 @@
+    """Weighting of a `NumpyTensorSpace` with constant 1."""
+
+    # Implement singleton pattern for efficiency in the default case
+    __instanc
```
This has to be fixed in conjunction with `matrix_representation` as mentioned above. Make an issue and look into it.

adler-j replied:
why not 0? the default is 0

adler-j replied:
You use a 3 dimensional array later.

Anyway, if we allow 2+, why not allow 1d as well, in which case this is an inner product.

kohr-h replied:
Seems to be some GH bug, this comment clearly ended up in the wrong place. Can you post a link?

kohr-h replied:
[Same here](https://github.com/odlgroup/odl/pull/861#discussion_r100929338)

adler-j replied:
It was related to `MatrixOperator`, I'll make a new comment.

adler-j replied:
You use a 3 dimensional array later.

Anyway, if we allow 2+, why not allow 1d as well, in which case this is an inner product.

adler-j replied:
This was regarding the axis argument in `MatrixOperator`, where you state that it must be >0, but the default is 0.

kohr-h replied:
Wrong place again, you need to hack it apparently.

adler-j replied:
The comment is w.r.t `MatrixOperator` where you assume the matrix is 2 or higher dimensional. I can easily see this working with 1 or even 0 dimensional arrays, so that condition is arbitrary.

The line "You use a 3 dimensional array later." refers to a comment that `MatrixOperator` input should be 2d, but you later use 3d. Also you later say that `axis` needs to be 1 or more, but the default is `0`.

adler-j commented on odl/space/pspace.py:
```diff
@@ -569,11 +569,17 @@ def __getitem__(self, indices):
+            return ProductSpace(*self.spaces[indices], field=self.field)
+        elif isinstance(indices, tuple):
+            if len(indices) > 1:
+                raise ValueError('too many indices: {}'.format(indices))
```
Another option is to pass these through further down, i.e.

```python
return ProductSpace(self.spaces[indices[0]][indices[1:]], field=self.field)
```

adler-j commented on odl/space/pspace.py:
```diff
@@ -848,7 +854,7 @@ def show(self, title=None, indices=None, **kwargs):
         odl.discr.lp_discr.DiscreteLpElement.show :
             Display of a discretized function
-        odl.space.base_ntuples.NtuplesBaseVector.show :
+        odl.space.base_tensors.GeneralizedTensor.show :
```
find replace `GeneralizedTensor`

adler-j commented on odl/space/space_utils.py:
```diff
@@ -43,13 +42,15 @@ def vector(array, dtype=None, impl='numpy'):
-    impl : string, optional
-        The backend to use. See `odl.space.entry_points.NTUPLES_IMPLS` and
-        `odl.space.entry_points.FN_IMPLS` for available options.
+    order : {'C', 'F'}, optional
```
'A'?

kohr-h replied:
Clearly missing

adler-j commented on odl/space/space_utils.py:
```diff
@@ -43,13 +42,15 @@ def vector(array, dtype=None, impl='numpy'):
-        `odl.space.entry_points.FN_IMPLS` for available options.
+    order : {'C', 'F'}, optional
+        Axis ordering of the data storage.
+    impl : str
```
optional

adler-j commented on odl/space/weighting.py:
```diff
@@ -686,25 +679,30 @@ def __init__(self, const, impl, exponent=2.0, dist_using_inner=False):
         if not np.isfinite(self.const):
             raise ValueError('`const` {} is invalid'.format(const))
 
+    def __hash__(self):
```
Prefer if `__hash__` is always right after `__eq__` since they are so strongly connected.

kohr-h commented:
Thanks *a lot* for the detailed review @adler-j, will try to get this done asap.

kohr-h commented on odl/test/largescale/space/tensor_space_slow_test.py:
```diff
@@ -39,119 +39,119 @@
 
-    x0 = np.arange(fn.size)
-    x = fn.element(x0)
+    x0 = np.arange(tspace.size)
```
Haven't looked much into this file, probably needs more love

kohr-h commented on doc/source/generate_doc.py:
```diff
@@ -63,26 +63,26 @@
 def import_submodules(package, name=None, recursive=True):
-    """ Import all submodules of a module, recursively, including subpackages
-    """
+    """Recursively import all submodules of ``package``."""
```
will check

kohr-h commented on doc/source/guide/glossary.rst:
```diff
@@ -22,7 +22,7 @@ Glossary
     discretization
         Structure to handle the mapping between abstract objects (e.g. functions) and concrete, finite realizations.
-        It encompasses an abstract `Set`, a finite data container (`NtuplesBaseVector` in general) and the mappings between them,         :term:`sampling` and :term:`interpolation`.
+        It encompasses an abstract `Set`, a finite data container (`GeneralizedTensor` in general) and the mappings between them,         :term:`sampling` and :term:`interpolation`.
```
Yes, needs to be replaced

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -32,97 +32,88 @@
-        else:
-            self.__order = str(order).upper()
+            if not fspace.has_field:
+                raise TypeError('`fspace.field` cannot be `None` for '
```
Right, maybe it was paranoia that made me hide the `None` behind a property. Will remove it.

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -32,97 +32,88 @@
+            if not fspace.has_field:
+                raise TypeError('`fspace.field` cannot be `None` for '
+                                '`linear=True`')
+            if not dspace.is_numeric:
```
Will try to come up with something reasonable. Or I'll check how often this is needed and if I can simply use the low-level ones instead.

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -382,57 +366,37 @@ def __init__(self, fset, partition, dspace, **kwargs):
+        - **Important:** if called on a point array, the points are
+          assumed to be sorted in ascending order in each dimension
+          for efficiency reasons.
+        - Nearest neighbor interpolation is the only scheme which works
```
Well, depends. Here I want to sell NN interpolation :-)

adler-j replied:
Then at least add a word on why the others *doesnt*

kohr-h replied:
Added a short remark

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -442,9 +406,7 @@ def nearest(arg, out=None):
-                self.grid.coord_vectors,
-                x.asarray().reshape(self.grid.shape, order=self.order),
-                variant=self.variant,
+                self.grid.coord_vectors, x.asarray(), variant=self.variant,
```
The idea from the beginning was that `_NearestInterpolator` and friends operate solely on arrays and do not have to know about ODL stuff. Admittedly, adding a `np.asarray` there would keep it that way. Why not.

kohr-h replied:
Actually, no changes required in the interpolators. Just removed the `.asarray()`

kohr-h commented on odl/discr/discretization.py:
```diff
@@ -241,40 +225,97 @@ def domain(self):
+    @property
+    def weighting(self):
+        """This space's weighting scheme."""
+        return getattr(self.dspace, 'weighting', None)
```
Okay.

kohr-h replied:
The `is_weighted` thing too? Feels like that one should rather return `False` in that case.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -1336,35 +1279,17 @@ def uniform_discr_fromdiscr(discr, min_pt=None, max_pt=None,
+        Desired minimum/maximum corners of the new space domain.
     shape : int or sequence of ints, optional
-        Number of samples per axis.
-    cell_sides : array-like, optional
```
Dunno, I suspect it has just been moved around. Will check.

kohr-h replied:
Ah, I moved all the deleted stuff into `kwargs` and just mention the pass-through below.

kohr-h commented on odl/discr/discretization.py:
```diff
@@ -400,163 +441,34 @@ def interpolation(self):
+        """Implement ``self **= p``."""
+        # Falls back to `LinearSpaceElement.__ipow__` if `self.tensor`
+        # has no own `__ipow__`. The fallback only works for integer `p`.
+        # TODO: do we even need this??
```
Yeah, this seems rather fishy. I'll see if there is a better way.

kohr-h replied:
A quick check reveals that this is actually needed. The thing is that `DiscretizedSpaceElement` inherits from `Tensor` (the abstract one) which has no own `__ipow__` and thus falls back to the `LinearSpaceElement` version which only works for integers.
Others are not needed since `__pow__` uses `__ipow__` and the rest has no modification in `Tensor` classes either.

kohr-h commented on odl/operator/default_ops.py:
```diff
@@ -1087,18 +1122,23 @@ def adjoint(self):
+
+    Implements::
+
+        ComplexEmbedding(x) == x + 1j * zero()
```
Will check. However, I believe I didn't really change anything here.

adler-j replied:
Worst case leave it for now, it is clearly understandable atm.

kohr-h commented on odl/solvers/functional/derivatives.py:
```diff
@@ -172,6 +172,7 @@ def __init__(self, functional, method='forward', step=None):
         functional : `Functional`
             The functional whose gradient should be computed. Its domain must
+            be a `TensorSpace`.
             be an `FnBase` space.
```
Looks like a rebasing artefact

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+from odl.util.utility import TYPE_MAP_R2C, TYPE_MAP_C2R
+
+
+__all__ = ('TensorSpace',)
```
I think the fact that we don't want users to initiate such an object directly is a strong reason for not exposing it at top level.

kohr-h replied:
Actually we shouldn't expose any of the classes that we don't want users to use. That includes includes *all* element type classes, also `LinearSpaceElement`.

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+    supported by this class, along with reductions based on arithmetic
+    or comparison, and element-wise mathematical functions ("ufuncs").
+
+    See `this Wikipedia article <https://en.wikipedia.org/wiki/Tensor>`_
```
I have a book I can cite. Perhaps there's also a more specific section at Wikipedia which is more in line with our case.

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+        return self.__is_real or self.__is_complex
+
+    @property
+    def is_real_space(self):
```
Hm, why not. Bloody awesome song :-)

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+
+        On the other hand, spaces are not unique:
+
+        >>> spc2 = odl.tensor_space((2, 3), dtype='uint64')  # same as spc
```
I actually changed them all on the GPU branch to simple shape, will do something similar here.

adler-j replied:
Bumping this

kohr-h replied:
Changed them in the other file only, will do the same here.

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+
+    @property
+    def ndim(self):
+        """Number of axes (=dimensions) of this space, also called "rank"."""
```
Well at least in tensor lingo, number of dimensions and rank are the same thing. However, when the dimensions have different interpretations, it sure needs additional thought. Most important is to keep compatibility with the `ProductSpace` version of the same thing. Luckily, those don't have a shape yet, so there's no ambiguity at the moment.

New issue I guess, and I remove the rank remark here.

adler-j replied:
*make the issue*

kohr-h replied:
#908

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+        return "{}({})".format(self.__class__.__name__,
+                               signature_string(posargs, optargs))
+
+    __str__ = __repr__
```
That's actually true, somewhat surprising to me (both assignment and override at class level, so no clear order). Should be changed then.

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+        """Name of the implementation back-end of this tensor."""
+        return self.space.impl
+
+    copy = LinearSpaceElement.copy
```
I'll check

kohr-h replied:
Along with `__init__` and `space`

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+        """Return ``self != other``."""
+        return not self == other
+
+    def __hash__(self):
```
I didn't think much about it here, thought I remembered another place where hashing of a numpy array was needed. But actually, this shouldn't be hashable, so I'll remove it.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -133,53 +175,137 @@ def out_dtype(self):
             return self.element_type(self, fcall)
 
+    def zero(self):
+        """Function mapping everything to zero.
```
anything is actually more correct

kohr-h commented on odl/space/fspace.py:
```diff
@@ -133,53 +175,137 @@ def out_dtype(self):
+            else:
+                # Need to go through an array to fill with the correct
+                # zero value for all dtypes
+                fill_value = np.ones(1, dtype=self.out_dtype)[0]
```
The dtype is not callable, so this won't work.

kohr-h replied:
And `1` is not the right input value for all dtypes.

adler-j replied:
```
In [3]: np.float32(1)
Out[3]: 1.0
```

what should it be in the other cases?

kohr-h replied:
Think of strings and stuff like that. Apart from that, the objects generated by `np.dtype()` cannot be called, so that's not an option.

adler-j replied:
Shouldn't this function throw an error if you ask for ones of a string?

kohr-h replied:
```python
>>> np.ones(3, dtype='U1')
array(['1', '1', '1'], 
      dtype='<U1')
>>> np.zeros(3, dtype='U1')
array(['', '', ''], 
      dtype='<U1')
```
Don't ask me why

kohr-h commented on odl/space/fspace.py:
```diff
@@ -133,53 +175,137 @@ def out_dtype(self):
 
-        return (isinstance(other, type(self)) and
-                isinstance(self, type(other)) and
+        return (isinstance(other, FunctionSpace) and
```
Yes, needs to change

kohr-h commented on odl/space/fspace.py:
```diff
@@ -998,10 +607,7 @@ def examples(self):
-            yield ('One', self.one())
-        except NotImplementedError:
-            pass
+        yield ('One', self.one())
```
We know that `one` exists?

adler-j replied:
What about non linearspace stuff?

kohr-h replied:
Exists there, too.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -1038,138 +644,380 @@ def _sphere_fun(x):
+
+        # Gaussian function
+        def _gaussian_fun(x):
+            if ndim == 1:
```
Old stuff, I can adapt it

adler-j commented on odl/discr/discretization.py:
```diff
@@ -320,10 +361,10 @@ def __getitem__(self, indices):
             The value(s) at the index (indices)
         """
-        return self.ntuple.__getitem__(indices)
+        return self.tensor.__getitem__(indices)
```
*make the issue before you forget it*

kohr-h replied:
> make the issue before you forget it

That's nontrivial either :-) Okay, I'll do it.

kohr-h replied:
Done, #907

kohr-h commented on doc/source/guide/glossary.rst:
```diff
@@ -11,7 +11,7 @@ Glossary
     array-like
         Any data structure which can be converted into a `numpy.ndarray` by the
-        `numpy.array` constructor. Includes all `NtuplesBaseVector` based classes.
+        `numpy.array` constructor. Includes all `Tensor` based classes.
```
Yeah scrap those shitty old names!

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+
+        Returns
+        -------
+        wrapper : `Tensor`
```
I think it's correct since the output wraps the input.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -133,53 +175,137 @@ def out_dtype(self):
+            elif is_valid_input_array(x, self.domain.ndim):
+                out_shape = out_shape_from_array(x)
+            else:
+                raise TypeError('invalid input type')
```
Works due to the magic in `__call__` (the Py function is wrapped, see below).

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,741 @@
+            self.__shape = tuple(safe_int_conv(s) for s in shape)
+        except TypeError:
+            self.__shape = (safe_int_conv(shape),)
+        if any(s < 0 for s in self.shape):
```
Actually, no. Numpy allows size 0, so we should, too. We even have a test that fails after I changed this.

adler-j replied:
Ok lets go with that then.

adler-j replied:
Make sure the doc mentions that we can deal with it.

adler-j commented:
What is the review status on this?

kohr-h commented:
> What is the review status on this?

I figured that sampling would need tensor-valued functions, so I implemented them. Now I'm in the middle of writing/fixing the tests for those.

adler-j commented:
Ok, just tell me when you need more help.

kohr-h commented:
I believe I have addressed all points of the last review. On top of that, we now can represent and discretize vector- and tensor-valued functions as discussed in #908. As usual, there are a lot of annoying corner cases, and the code is more complex than I wished, but with all the flexibility (broadcasting mostly) per component, and the usual issues in 1d, things get a bit messy. 

But all tests pass now again, so I guess it's time for a new review.

kohr-h commented:
Oh, and I should mention that I still haven't touched the large-scale tests. I'm gonna do that when everything else is finished.

kohr-h commented on odl/operator/default_ops.py:
```diff
@@ -1087,18 +1123,23 @@ def adjoint(self):
-        space : `FnBase`
-            Space which real part should be taken, needs to implement
+        space : `TensorSpace`
+            Space whose into which should be embedded, needs to implement
```
oops

kohr-h commented on odl/space/fspace.py:
```diff
@@ -15,581 +15,233 @@
+
+
+def _check_out_arg(func):
+    """Check if of ``func`` has an (optional) ``out`` argument.
```
Argh!

kohr-h commented on odl/space/fspace.py:
```diff
@@ -15,581 +15,233 @@
+    except ValueError:
+        out[:] = result
+    else:
+        out[:] = reshaped
```
This was the simplest way to handle all annoying cases with extra dimensions. Not sure if this should be handled more explicitly.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -636,88 +317,313 @@ def element(self, fcall=None, vectorized=True):
         vectorized : bool, optional
-            Whether ``fcall`` supports vectorized evaluation.
+            If ``True``, assume that ``fcall`` supports vectorized
+            evaluation. For ``False``, , the function is decorated with a
```
double comma

kohr-h commented on odl/space/fspace.py:
```diff
@@ -636,88 +317,313 @@ def element(self, fcall=None, vectorized=True):
             else:
-                out.fill(1)
+                # Need to go through an array to fill with the correct
+                # zero value for all dtypes
```
one

kohr-h commented on odl/space/fspace.py:
```diff
@@ -1150,26 +943,337 @@ def __init__(self, fspace, fcall):
+                    out = broadcast_to(out, out_shape)
+
+            elif self.space.tensor_valued:
+                # TODO: fix case when such a function is evaluated at a
```
can be removed

kohr-h commented on odl/tomo/backends/astra_cpu.py:
```diff
@@ -66,18 +66,18 @@ def astra_cpu_forward_projector(vol_data, geometry, proj_space, out=None):
-                        'instance of `NumpyNtuples`'
-                        ''.format(vol_data.space.dspace))
+        raise TypeError('`vol_data` must be a `numpy.ndarray` based, '
+                        "container got `impl` {!r}"
```
misplaced comma

kohr-h commented on odl/operator/default_ops.py:
```diff
@@ -1087,18 +1123,23 @@ def adjoint(self):
             ``space.complex_space``.
         scalar : ``space.complex_space.field`` element, optional
-            Scalar which the incomming vectors should be multiplied by in order
+            Scalar to be multiplied with incoming vectors in order
```
Why do we have this parameter? IMO it's not the responsibility of this operator to multiply with anything, that should be done separately.

adler-j commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,828 @@
+        >>> space[0, 1:]
+        rn((2, 4))
+        """
+        if isinstance(indices, Integral):
```
Cant this whole thing be written:

```
tuple(np.asarray(self.shape)[indices])
```

If not, why?

adler-j commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,828 @@
+        """
+        return TensorSpaceUfuncs(self)
+
+    def show(self, title=None, method='', indices=None, force_show=False,
```
Has this been updated?

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -32,97 +32,89 @@
-            discretized object. Its `NtuplesBase.size` must be equal
-            to the total number of grid points.
         linear : bool, optional
+            discretized object. Its `TensorSpace.shape` must be equal
```
Wrong rebase as it seems, good spot.

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -176,127 +161,133 @@ class PointCollocation(FunctionSetMapping):
+
+    def _call(self, func, out=None, **kwargs):
+        """Return ``self(func[, out, **kwargs])``."""
         mesh = self.grid.meshgrid
```
Hm, I guess that would have to go into the class docstring. I'll check.

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -382,57 +362,38 @@ def __init__(self, fset, partition, dspace, **kwargs):
-        arithmetic operations on the values.
-        """
+        """Return ``self(x[, out])``."""
         # TODO: pass reasonable options on to the interpolator
```
Dunno, this is an old comment. No idea what I meant exactly.

kohr-h commented on odl/discr/discr_mappings.py:
```diff
@@ -676,23 +595,30 @@ def __repr__(self):
-            sep = ', '
+        posargs = [self.range, self.grid, self.domain, schemes]
+
+        nontriv_var = None
```
Ok, thought it was obvious :-P

kohr-h commented on odl/discr/discretization.py:
```diff
@@ -153,15 +138,15 @@ def dspace_type(self):
     @property
     def sampling(self):
-        """Operator mapping a `uspace` element to an n-tuple."""
+        """Operator mapping a `uspace` element to a tensor."""
```
yes

kohr-h commented on odl/discr/discretization.py:
```diff
@@ -241,40 +225,94 @@ def domain(self):
-
-        NtuplesBaseVector.__init__(self, space)
-        self.__ntuple = ntuple
+        super().__init__(space)
```
Oh, I actually didn't want to change that one.

But anyway, the non-speed-critical ones should use `super`, simply because it allows dependency injection, something that can be very useful when subclassing. An we should change the speed-critical ones also as soon as Python 2 is dropped.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -116,51 +108,43 @@ def __init__(self, fspace, partition, dspace, exponent=2.0,
-                raise ValueError("`interp` type '{}' not understood"
-                                 "".format(interp_in))
-            self.__interp_by_axis = [interp] * partition.ndim
+            interp + ''
```
ok

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -357,35 +348,13 @@ def element(self, inp=None, **kwargs):
-            return self.element_type(self, self.dspace.element(arr))
-
-    def _astype(self, dtype):
+            return self.element_type(self, self.dspace.element(inp))
```
Yeah, the implementation becomes so obvious..

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -404,8 +373,7 @@ def _inner(self, x, y):
             # TODO: implement without copying x
-            func_list = _scaling_func_list(bdry_fracs)
-
+            func_list = _scaling_func_list(bdry_fracs, exponent=1.0)
```
Because exponent is only used for norm if something to the p-th power should be integrated. It's added here because it was an optional parameter in the helper before, which didn't really make sense.

kohr-h commented on odl/discr/lp_discr.py:
```diff
@@ -894,43 +788,54 @@ def uniform_discr_frompartition(partition, exponent=2.0, interp='nearest',
-            'none' : no weighting
+            ``'none'`` : no weighting
+
+            `Weighting` : Use existing weighting as-is (not checked)
```
good question

kohr-h commented on odl/operator/default_ops.py:
```diff
@@ -431,7 +442,9 @@ def __str__(self):
-        ``InnerProductOperator(y)(x) <==> y.inner(x)``
+    Implements::
+
+        InnerProductOperator(y)(x) <==> y.inner(x)
```
ok

kohr-h commented on odl/operator/tensor_ops.py:
```diff
@@ -853,59 +888,650 @@ def inverse(self):
+        return repr(self)
+
+
+class FlatteningOperator(Operator):
```
Hm, okay, I'll think of a name. But let's do that in a separate PR.

adler-j replied:
Make an issue :D

kohr-h commented on odl/set/space.py:
```diff
@@ -866,20 +892,8 @@ def __ne__(self, other):
-        """
-        return repr(self.space) + 'Element'
+        """Return ``str(self)``."""
+        return repr(self)
```
Nowhere, but the default is to just print the class name, which I guess is the most sensible thing to do here.

kohr-h commented:
> How are tensor valued discretized functions coming along?

In principle it's clear what to do I'd say, but I would rather not make it part of this PR - it's already large enough. I did the continuous tensor-valued functions since I had no choice: the interpolator created in an interpolation operator is such a guy.
Of course, it's a natural follow-up issue.

kohr-h commented:
Rebased on master, will look into the details asap.

adler-j commented:
> In principle it's clear what to do I'd say, but I would rather not make it part of this PR - it's already large enough. I did the continuous tensor-valued functions since I had no choice: the interpolator created in an interpolation operator is such a guy.
Of course, it's a natural follow-up issue.

Make an issue then :)

kohr-h commented:
> In principle it's clear what to do I'd say, but I would rather not make it part of this PR - it's already large enough. I did the continuous tensor-valued functions since I had no choice: the interpolator created in an interpolation operator is such a guy.
>     Of course, it's a natural follow-up issue.
> 
> Make an issue then :)

Well, I'd say the issue for that already exists (along with some discussion about how to do this): #908

adler-j commented on odl/test/space/npy_tensors_test.py:
```diff
@@ -0,0 +1,1309 @@
+    """Check if space indexing works as expected."""
+    space = odl.tensor_space((2, 3, 4), dtype=complex, exponent=1, weighting=2)
+
+    # Ellipsis not supported
```
I mean why not with my proposed fix above

kohr-h replied:
The thing is that there's not really anything interesting to index here. For instance, `space[:2, :]` and `space[1:3, :]` are exactly the same since there's no information attached to the points along the axes. That's different for grids and also product spaces.
Here a utility that condenses an index expression into a new shape would do the deal.

adler-j replied:
For me it is more a manner of consistency, e.g. if  `vector[STUFF]` works, then `space[STUFF]` should also work, no?

adler-j commented on odl/test/space/npy_tensors_test.py:
```diff
@@ -0,0 +1,1309 @@
+    assert float(element) == float(value)
+    assert complex(element) == complex(value)
+    if PYTHON2:
+        long = eval('long')
```
wat?

kohr-h replied:
hehe

adler-j commented on odl/test/space/npy_tensors_test.py:
```diff
@@ -0,0 +1,1309 @@
+    assert np.array_equal(x_arr, np.ones(x.shape))
+    x_as_arr = np.asarray(x)
+    assert np.array_equal(x_as_arr, np.ones(x.shape))
+
```
We should perhaps try `asanyarray` here aswell

adler-j commented on odl/test/space/pspace_test.py:
```diff
@@ -491,6 +491,10 @@ def test_getitem_single():
         H[-3]
         H[2]
 
+    assert H[(1,)] == r2
```
I guess we should add more new tests here?

adler-j commented on odl/util/__init__.py:
```diff
@@ -6,7 +6,7 @@
 # obtain one at https://mozilla.org/MPL/2.0/.
 
-"""Utility library for ODL, only for internal use."""
+"""Utility library for ODL, mainly for internal use."""
```
Perhaps we should make an effort to fix that at some point

kohr-h replied:
Fix in what sense? More clearly separate between internal utils and user utils?
In some sense we do a bit of that by choosing which names to expose to the `odl.util` namespace.

adler-j replied:
I agree that this is an OK way to go for now, but later on we should perhaps add odl.internal

adler-j commented on odl/util/ufuncs.py:
```diff
@@ -68,6 +69,9 @@
 """.format(name)
     UFUNCS.append((name, n_in, n_out, doc))
 
+# TODO: add the following reductions (to the CUDA implementation):
```
We should still have them available here, no?

adler-j commented on odl/util/ufuncs.py:
```diff
@@ -136,70 +140,94 @@ def wrapper(self, x2, out=None):
 
-    def wrapper(self):
-        return wrapped(self.vector)
+    def wrapper(self, axis=None, out=None, keepdims=False, **kwargs):
```
Remove arguments to ...

kohr-h replied:
Sorry what?

adler-j replied:
Frankly cant remember

adler-j commented on odl/util/ufuncs.py:
```diff
@@ -224,50 +252,110 @@ def wrapper(self, x2, out=None):
 
+# For DiscreteLP, basically the ufunc mechanism can be propagated from its
+# `tensor` attribute. Sometimes, reshaping is required.
+# TODO: update this!
```
Is this broken? I guess it deserves a fix in that case

kohr-h replied:
No, just the comment is outdated.

adler-j commented on odl/util/utility.py:
```diff
@@ -49,7 +49,12 @@ def array1d_repr(array, nprint=6):
+        if array.size == 1:
+            return repr(np.array(array, ndmin=1)[0])
+        else:
+            return ''
```
Is this truly the right result in this case? Shouldn't it be `np.array()` or something like that?

kohr-h replied:
I redid the whole thing, using `np.array2string`.

kohr-h commented:
> Ok, I've just done the largest review of my whole life.

Monster PR deserves monster review :-)



> Overall, I can say that this is an exceptionally good PR, most things are very minor. With that said, some of the comments are stuff that should be fixed so some more work is needed. With that said, those are "point fixes", and shouldn't require another thorough review.
> 
> Very well done!

Cool, so I'll bring this up to speed regarding the comments, and perhaps when there are significant changes I'll ask for a per-commit review. Thanks so far.

Btw, I'm waiting for a review on my ufuncs branch from the Theano guys, and when that one is in the gpuarray branch will be significantly smaller. I'll probably just remove the ufunc implementation from that PR and assume it's included upstream (or import with safeguard).

kohr-h commented on odl/space/base_tensors.py:
```diff
@@ -0,0 +1,819 @@
+    def examples(self):
+        """Return example random vectors."""
+        # Always return the same numbers
+        rand_state = np.random.get_state()
```
I made an issue #999

kohr-h commented on odl/space/fspace.py:
```diff
@@ -627,88 +308,310 @@ def element(self, fcall=None, vectorized=True):
+                if out is None:
+                    # Out-of-place evaluation
+
+                    # Collect results of member functions into a list
```
I don't think I can make it simpler or more readable since there are really a bunch of nasty corner cases to be handled. But I'll add a more explanatory docstring to the wrapper.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -899,72 +750,67 @@ def ipow_posint(x, n):
-        x_call_oop = x._call_out_of_place
+    def _realpart(self, f):
+        """Function returning the real part of the result from ``f``."""
+        # Avoid infinite recursions by making a copy of the function
```
Happened sometime along the way, but with this implementation the problem doesn't occur any more. I'll remove this extra stuff.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -1141,26 +931,334 @@ def __init__(self, fspace, fcall):
+
+            element-like: must be a castable to a domain element
+
+            meshgrid: length must be ``space.ndim``, and the arrays must
```
Added some doc

kohr-h commented on odl/space/fspace.py:
```diff
@@ -1141,26 +931,334 @@ def __init__(self, fspace, fcall):
+        else:
+            self._call_in_place(x, out=out, **kwargs)
+
+    def __call__(self, x, out=None, **kwargs):
```
Agreed. The writing of this  was also mostly incremental, fixing one failing test at a time.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -1141,26 +931,334 @@ def __init__(self, fspace, fcall):
+                elif ndim == 1 and out.shape == (1,) + out_shape:
+                    out = out.reshape(out_shape)
+
+                if out_shape not in ((), (1,)) and out.shape != out_shape:
```
Actually only the `()` case should be filtered out here, since we want
```python
>>> fspace = odl.FunctionSpace(odl.IntervalProd(0, 1), out_dtype=(float, (1,)))
>>> func = fspace.element(lambda x: x)
>>> func(0)
array([ 0.])
```
Fixing

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+
+            float: Weighting by a constant
+
+            array-like: Pointwise weighting by an array
```
Not yet, should go to the TODO list.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+        >>> same_space = odl.NumpyTensorSpace((2, 3), dtype='float64')
+        >>> same_space == space_a
+        True
+        """
```
Well, for me the  guideline is 
```python
>>> np.array_equal([3], [[3]])
False
```
The two spaces clearly differ in the number of axes.

kohr-h commented on odl/space/fspace.py:
```diff
@@ -627,88 +308,310 @@ def element(self, fcall=None, vectorized=True):
+                    'an array-like of callables whose shape matches '
+                    '`out_shape` {}'.format(self.out_shape))
+
+            fcalls = np.array(fcall, dtype=object, ndmin=1).ravel().tolist()
```
Example added

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+        >>> space[0, 1:]
+        rn((2, 4))
+        """
+        if isinstance(indices, Integral):
```
Actually now that I look at this again I really wonder if we need this at all. I don't really see a good use case.

The actually useful functionality is indexing `byaxis`.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+        Returns
+        -------
+        values : `NumpyTensorSpace.dtype` or `NumpyTensor`
+            The value(s) at the given indices. Note that the returned
```
Not when `indices` is a list, see example below. Clarified in the text now.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+                # To preserve the array view for non-contiguous slices,
+                # we need to use 'A' for the space in that case.
+                out_spc_order = 'A'
+            space = space_constructor(
```
Not really, at least not as long as `space.__getitem__` really captures all the relevant cases. There's much more going on with Numpy indexing that would be very tedious to re-implement. I like this lazy approach much more, since `arr[indices]` tells us everything we need to know.

As I wrote above, I really see rather little use for `space.__getitem__` unless it really does what you write here.

adler-j replied:
Well why don't we simply use numpy for indexing then, i.e. wrap in a numpy array, perform indexing, unwrap?

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+             [-2.0, 1.0, -2.0]]
+        )
+
+        Be aware of unsafe casts and over-/underflows, there
```
Yeah, we had this before and I kept it, but I also don't think it's super important.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+    else:
+        ravel_order = 'C'
+
+    x1_arr = x1.data.ravel(order=ravel_order)
```
Hm, yes. I changed it to not ravel if stuff is not contigouous in the same way. The fallbacks will handle that now. Not sure if there are proper tests for non-contiguous stuff, need to check.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+    # BLAS dot or nrm2
+    xp = np.abs(x.data.ravel(order=x.order))
+    if p == float('inf'):
+        xp *= w.ravel(order=x.order)  # w is a plain NumPy array
```
Sure, but that's a case for the time when that feature is implemented.

kohr-h commented on odl/space/npy_tensors.py:
```diff
@@ -0,0 +1,1992 @@
+                       x1.data.ravel(order=x1.order))
+
+
+# TODO: implement intermediate weighting schemes with arrays that are
```
#1001

kohr-h commented on odl/space/pspace.py:
```diff
@@ -556,16 +538,92 @@ def __hash__(self):
         return hash((type(self), self.spaces, self.weighting))
 
     def __getitem__(self, indices):
```
Which would also make them consistent with how `FunctionSpace` handles it.

kohr-h commented on odl/space/pspace.py:
```diff
@@ -556,16 +538,92 @@ def __hash__(self):
+        >>> pspace2[:-1, 0]
+        ProductSpace(rn(2), 2)
+        """
         if isinstance(indices, Integral):
```
I wouldn't really know how without actually doing the recursion. I'll make a utility when we [know we need it](https://github.com/odlgroup/odl/pull/861/files/e48ce31bcf2d10a6334996800c5a0c3804dd38c4#diff-f6ddef3abbfa8c9763285b791e19d177).

kohr-h commented on odl/space/weighting.py:
```diff
@@ -466,35 +420,29 @@ def equiv(self, other):
     def repr_part(self):
         """Return a string usable in a space's ``__repr__`` method."""
-        if self.matrix_issparse:
+        if scipy.sparse.isspmatrix(self.matrix):
```
Seemed like namespace pollution for such a trivial functionality.

kohr-h commented on odl/space/weighting.py:
```diff
@@ -598,18 +530,18 @@ def __eq__(self, other):
         """Return ``hash(self)``."""
         # TODO: Better hash for array?
-        return super().__hash__() ^ hash(self.array.tostring())
+        return super().__hash__() ^ hash(self.array.tobytes())
```
Probably we need to consider this non-hashable, no?

kohr-h commented on odl/test/discr/lp_discr_test.py:
```diff
@@ -7,213 +7,185 @@
+    assert discr.is_real_space
+    assert discr.dspace.exponent == 2.0
+    assert discr.dtype == discr.dspace.default_dtype(odl.RealNumbers())
+    assert all_equal(discr.min_pt, np.array([0]))
```
Not really since it's just pass-through from `partition`.

kohr-h commented on odl/test/space/fspace_test.py:
```diff
@@ -110,683 +49,1111 @@ def _meshgrid(domain, shape):
+    assert fspace_c.real_out_dtype == float
+    assert fspace_c.complex_out_dtype == complex
+    assert fspace_s.out_dtype == np.dtype('U1')
+    assert fspace_s.real_out_dtype is None
```
Hm yeah, good point

kohr-h commented on odl/test/space/fspace_test.py:
```diff
@@ -110,683 +49,1111 @@ def _meshgrid(domain, shape):
+    fspace_r = FunctionSpace(intv, out_dtype=float)
+    fspace_c = FunctionSpace(intv, out_dtype=complex)
+    fspace_intv2 = FunctionSpace(intv2)
+    fspace_vec = FunctionSpace(intv, out_dtype=(float, (2,)))
```
To test exactly what? For me those are two different things.

kohr-h commented on odl/test/space/fspace_test.py:
```diff
@@ -110,683 +49,1111 @@ def _meshgrid(domain, shape):
+        f_oop = fspace.element(func_nd_oop)
+        f_ip = fspace.element(func_nd_ip)
+        f_dual = fspace.element(func_nd_dual)
+    elif ndim == 1:
```
True, can't think of a good reason.

kohr-h replied:
Second thought, defining a tensor-valued function with non-square and non-trivial shape requires at least `(2, 3)`, which makes it quite a hassle to do for all cases relevant in these tests. So I took shape `(2,)` for the comprehensive tests and added just a few cases for shape `(2, 3)`.

adler-j replied:
Sounds like a good compromise

adler-j commented:
Can you re-open this for a new review, if you want one, else we *really* want to go ahead here.

kohr-h commented:
After thinking about it some more, I think the following way of indexing spaces is most reasonable:

- `NumpyTensorSpace`:
  - `__getitem__`: Numpy indexing, should wrap an array indexed in the same way
  - `byaxis`: index axes of the tensors
- `FunctionSpace`:
  - `__getitem__`: Numpy indexing for the `out_shape` part, i.e., for tensor-valued functions. Raise an error for scalar-valued functions.
  - `byaxis`: index the domain per axis
- `DiscreteLp`: same as `FunctionSpace`

Opinions?

kohr-h commented:
The page for this PR has become unworkable once again, opening a new one.
